### ISSUE TITLE
Possibly rename Floki.apply_transformation to Floki.map

### ISSUE THREAD

**Issue by aphillipo** [philss/floki#118]
Hi @philss! Thanks again for Floki... 

I have a quick question about apply_transformation - in my head I was looking for Floki.map rather than Floki.apply_transformation.  I wonder if you think the current naming is correct?

In my head Enum.map means iterate over all all items in an Enumerable; Floki.map means the same for all elements in a Floki html_tree?

I'm going to use apply_transformation now because I need to blur out some text and scramble the letters before payment unblurs them with unscrambled text.

I'm happy to rename it and add some examples if you want a pull request?  Thanks!

**Comment by philss** [comment]
Hi @aphillipo! 

I agree with you. I'm not that happy with the current name, but I did not realise that this is basically a map. I would love if you rename it!

I also think in the future we could add an implementation of the `Enumerable` protocol to the `Floki.HTMLTree` struct. This would make easy to traverse and apply some transformations.

Thank you!

**Comment by aphillipo** [comment]
Oh!  That is a great idea @philss - would all of Enum then work with Floki?

**Comment by aphillipo** [comment]
I'm looking at implementing Enumerable for ```Floki.HTMLTree``` now.  It'll be good for to level up my Elixir ability!  Thanks!

**Comment by philss** [comment]
Yes, I never did such thing (implementing a protocol for a complex struct), but I think it works! Ping me if you need help with something. Thanks!

**Comment by aphillipo** [comment]
I'm not certain I don't prefer the original tuples/lists of the html_tree type :-/

Something like this get's us 99% of the way there, but I'm not sure you can implement Enumerable unless you have a defstruct type.

https://github.com/rrrene/html_sanitize_ex/blob/master/lib/html_sanitize_ex/traverser.ex

A little confused by the ordering of Floki.HTMLTree:

1) Root Node looks to always be 1, but also it seems possible to have multiple root nodes? Go through these first. floki_tree[1], floki_tree[2] etc.

2) Each time child nodes has values [h|t] add the head to the accumulator with {:cont} as the Enumerable action and recurse on the tail.  Will probably get the actual nodes here for all children and pass them in.

3) Can probably use the default implementations for count and member (for now) because they just use the reducer.  Presumably the match for member would be an exact match on the particular struct.

4) Feel like writing this out have helped a bit.  Liking the HTMLTree a bit more now.  Let me try again!

**Comment by philss** [comment]
Hi @aphillipo!

Answering the points:

> 1. Root Node looks to always be 1, but also it seems possible to have multiple root nodes? Go through these first. floki_tree[1], floki_tree[2] etc.

It is possible to have more root nodes when you have, for example this kind of document fragment:

```html
<div></div>
<div></div>
```

It does not have a parent, and has more than one root. This way, the two divs are root elements.


> 2. Each time child nodes has child nodes [h|t] add the head to the accumulator with {:cont} as the Enumerable action and recurse on the tail. Will probably get the actual nodes here for all children and pass them in.

I'm not sure if I understood correctly, but I think you can get the `node_ids` from the `HTMLTree struct` as reference. If I remember correctly, It will always have the correct nodes order, but reversed .



> 3. Can probably use the default implementations for count and member (for now) because they just use the reducer. Presumably the match for member would be an exact match on the particular struct.

If you use the `node_ids`, you will probably get the `count` right. Not sure how `member` would work.


> 4. Feel like writing this out have helped a bit. Liking the HTMLTree a bit more now. Let me try again!

Nice! I think I should promote more about this struct. It enables a lot of cool things, like updating the tree or easily get nodes by its types, without having to traverse the tree using tuples.

**Comment by philss** [comment]
I did some work based on our conversation. I came up with this PR merged: https://github.com/philss/floki/pull/127 and also this one that is open #128.

Please check and tell me what you think.

Thanks!

**Comment by aphillipo** [comment]
This is brilliant!  And very good timing for me!  I'm creating micro payments platform for articles and I need to be able to give a sensible excerpt from a HTML structure.  Obviously making something a certain length with tag's etc. isn't simple without something like this!  I'll implement my excerpt/filter bad tags stuff and get back to you with any issues!

**Comment by philss** [comment]
@aphillipo Nice, thank you! 😃 
I will merge this PR and close the issue. Please reopen it if you think it's needed. I will release a version soon with this change.

And one more thing: you said you need to excerpt bad tags. Maybe `Floki.filter_out/2` can help you.
